include/sys/trace.h

```diff
@@ -65,4 +72,87 @@
 #define	TRACE_INCLUDE_FILE trace
 #include <trace/define_trace.h>
 
-#endif /* _KERNEL && HAVE_DECLARE_EVENT_CLASS */
+#else /* HAVE_DECLARE_EVENT_CLASS */
+
+#define	DTRACE_PROBE(name) \
+	trace_zfs_##name()
+
+#define	DTRACE_PROBE1(name, t1, arg1) \
+	trace_zfs_##name((uintptr_t)(arg1))
+
+#define	DTRACE_PROBE2(name, t1, arg1, t2, arg2) \
+	trace_zfs_##name((uintptr_t)(arg1), (uintptr_t)(arg2))
+
+#define	DTRACE_PROBE3(name, t1, arg1, t2, arg2, t3, arg3) \
+	trace_zfs_##name((uintptr_t)(arg1), (uintptr_t)(arg2), \
+	(uintptr_t)(arg3))
+
+#define	DTRACE_PROBE4(name, t1, arg1, t2, arg2, t3, arg3, t4, arg4) \
+	trace_zfs_##name((uintptr_t)(arg1), (uintptr_t)(arg2), \
+	(uintptr_t)(arg3), (uintptr_t)(arg4))
```

Comment:
Not a big deal, but: The difference from HAVE_DECLARE_EVENT_CLASS is that here we are casting the args to uintptr_t.  Could we do the same with HAVE_DECLARE_EVENT_CLASS so that we don't have to duplicate this code?

Reply:
With tracepoints based probes the arguments type is used in the functions defined by the tracepoint DECLARE_EVENT_CLASS.  So the uinptr_t casts lead to compilation errors:

In file included from ./include/zfs/sys/zfs_context.h:64:0,
                 from ./include/zfs/sys/spa.h:37,
                 from fs/zfs/zfs/arc.c:278:
fs/zfs/zfs/arc.c: In function ‘arc_evict_hdr’:
./include/zfs/sys/trace.h:30:19: warning: passing argument 1 of ‘trace_zfs_arc__delete’ makes pointer from integer without a cast [-Wint-conversion]
  trace_zfs_##name((uintptr_t)(arg1))
                   ^
fs/zfs/zfs/arc.c:3955:3: note: in expansion of macro ‘DTRACE_PROBE1’
   DTRACE_PROBE1(arc__delete, arc_buf_hdr_t *, hdr);
   ^~~~~~~~~~~~~
In file included from ./include/trace/syscall.h:5:0,
                 from ./include/linux/syscalls.h:84,
                 from ./include/zfs/spl/sys/vnode.h:29,
                 from ./include/zfs/spl/sys/kobj.h:28,
                 from ./include/zfs/sys/zfs_context.h:45,
                 from ./include/zfs/sys/spa.h:37,
                 from fs/zfs/zfs/arc.c:278:
./include/linux/tracepoint.h:233:21: note: expected ‘arc_buf_hdr_t * {aka struct arc_buf_hdr *}’ but argument is of type ‘long unsigned int’
  static inline void trace_##name(proto)    \
                     ^
./include/linux/tracepoint.h:396:2: note: in expansion of macro ‘__DECLARE_TRACE’
  __DECLARE_TRACE(name, PARAMS(proto), PARAMS(args),  \
  ^~~~~~~~~~~~~~~
./include/linux/tracepoint.h:521:2: note: in expansion of macro ‘DECLARE_TRACE’
  DECLARE_TRACE(name, PARAMS(proto), PARAMS(args))
  ^~~~~~~~~~~~~
./include/zfs/sys/trace_arc.h:103:1: note: in expansion of macro ‘DEFINE_EVENT’
 DEFINE_EVENT(zfs_arc_buf_hdr_class, name, \
 ^~~~~~~~~~~~
./include/zfs/sys/trace_arc.h:109:1: note: in expansion of macro ‘DEFINE_ARC_BUF_HDR_EVENT’
 DEFINE_ARC_BUF_HDR_EVENT(zfs_arc__delete);
 ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from ./include/zfs/sys/zfs_context.h:64:0,
                 from ./include/zfs/sys/spa.h:37,
                 from fs/zfs/zfs/arc.c:278: